### ISSUE TITLE
Unable to start server on host when using 64bit kernel on Raspbian (Or any 32-bit userland)

### ISSUE THREAD

**Issue by grebeplusplus** [microsoft/vscode-remote-release#2947]
- VSCode Version: (From About)
Version: 1.44.2 (user setup)
Commit: ff915844119ce9485abfe8aa9076ec76b5300ddd
Date: 2020-04-16T16:36:23.138Z
Electron: 7.1.11
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.18363
- VSCode Extensions: ms-vscode-remote.remote-ssh, ms-vscode-remote.remote-ssh-nightly
- Local OS Version: Microsoft Windows [Version 10.0.18363.815]
- Remote Hardware: RaspberryPi 4 Model B 2GB
- Remote OS Version: Linux 4.19.97-v8+ #1294 SMP PREEMPT Thu Jan 30 13:27:08 GMT 2020 aarch64 GNU/Linux
- Remote Extension/Connection Type: SSH

Steps to Reproduce:

1. Install Raspbian on RaspberryPi (2020-02-13-raspbian-buster-lite)
2. Install 64-bit kernel on RaspberryPi (rpi-update)
3. set `arm_64bit=1` under section `[pi4]` in `/boot/config.txt`
3. Install VSCode (1.44.2) on Windows PC
4. Configure public key SSH using `%USERPROFILE%\.ssh\config`
5. Install VSCode Remote Development (ms-vscode-remote.vscode-remote-extensionpack)
6. Connect to existing configuration host from config file for RaspberryPi
7. Observe failure talking about `.../server.sh: 12: .../server.sh: .../node: not found

Issue:
node reports as `not found` despite file existing in the file system. This is not due to node not being found but a dependency of node: the GNU C runtime which is dynamically linked: ld-linux-aarch64.so.

Root cause:
Raspbian runs a 32-bit userland (armhf) even on a 64-bit kernel (aarch64). vscode's installer scripts incorrectly determine the architecture of the platform without ensuring that the proper runtime support is installed.

Official Raspbian apt repos do *NOT* support 64bit packages at this time
> N: Skipping acquire of configured file 'main/binary-arm64/Packages' as repository 'http://archive.raspberrypi.org/debian buster InRelease' doesn't support architecture 'arm64'

What's more these libs cannot be added on simply by adding a foreign architecture using dpkg.
>The following packages have unmet dependencies:
> gcc-8-arm-linux-gnueabi:arm64 : Depends: cpp-8-arm-linux-gnueabi:arm64 (= 8.3.0-2cross1) but it is not going to be installed
>                                 Depends: libcc1-0:arm64 (>= 8) but it is not going to be installed
>                                 Depends: binutils-arm-linux-gnueabi:arm64 (>= 2.31.1) but it is not going to be installed
>                                 Depends: libgcc-8-dev-armel-cross:arm64 (>= 8.3.0-2cross1) but it is not installable
>                                 Depends: libc6:arm64 (>= 2.17) but it is not going to be installed
>                                 Depends: libgcc1:arm64 (>= 1:3.0) but it is not going to be installed
>                                 Depends: libgmp10:arm64 (>= 2:5.0.1~) but it is not going to be installed
>                                 Depends: libisl19:arm64 (>= 0.15) but it is not going to be installed
>                                 Depends: libmpc3:arm64 but it is not going to be installed
>                                 Depends: libmpfr6:arm64 (>= 3.1.3) but it is not going to be installed
>                                 Depends: libstdc++6:arm64 (>= 5) but it is not going to be installed
>                                 Depends: zlib1g:arm64 (>= 1:1.1.4) but it is not going to be installed
>                                 Recommends: libc6-dev-armel-cross:arm64 (>= 2.13-5)

Diagnostics:
~ $ .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/server.sh
> .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/server.sh: 12: .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/server.sh: .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/node: not found

~ $ ls .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/node
> .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/node

~ $ .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/node
> -bash: .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/node: No such file or directory

~ $ file .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/node
> .vscode-server/bin/ff915844119ce9485abfe8aa9076ec76b5300ddd/node: ELF 64-bit LSB executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=563924f2479df10436f7cbf69ec932d557fb399f, with debug_info, not stripped

~ $ ls  /lib/ld-linux-aarch64.so.1
> ls: cannot access '/lib/ld-linux-aarch64.so.1': No such file or directory

~ $ ls /lib/ld-linux-*
> /lib/ld-linux-armhf.so.3

~ $ uname -a
> Linux raspi-15E4C04D2 4.19.97-v8+ #1294 SMP PREEMPT Thu Jan 30 13:27:08 GMT 2020 aarch64 GNU/Linux

~ $ dpkg --print-architecture
> armhf

Expectation:
vscode should respect the dpkg architecture on Debian derivative platforms to ensure comparability 

Possible Workaround:
* Install a 64-bit guest OS version of Debian in a container.

Comments:
@roblourens and @siriuslysam got very close to the right answer when they were debugging the setup. The issue at hand seems more one of poor debug output relating to hidden/assumed dependancies on glibc being "always" present on Linux for whatever the kernel's architecture is.

**Comment by roblourens** [comment]
Yeah currently I just look at the aarch64 in uname and assume the 64-bit version is what we want. So the right solution is just to also check `dpkg` and install the 32 bit version in this case right?

**Comment by grebeplusplus** [comment]
For Debian that should work.

However you still have an unmet dependency in the case of the bug complaining that the system does not have glibc installed at all (for some reason I thought that was bug 443 but I seem to be mistaken I wish I could find the issue number that it was but it was late last night).

I think you may end up needing to either statically link your C runtime (which is allowed under an exception to the GPL) or triggering an install using the system package manager to resolve your dependency.

As an aside... I'm more of a Windows guy than a Linux guy... do we know who we can report that "not found" error message that doesn't say what wasn't found to? It's really not helpful and leads people down bad debugging rabbit holes.

**Comment by roblourens** [comment]
I guess to node, like https://github.com/nodejs/node, but I don't know, it may be upstream of them.

> However you still have an unmet dependency in the case of the bug complaining that the system does not have glibc installed at all (for some reason I thought that was bug 443 but I seem to be mistaken I wish I could find the issue number that it was but it was late last night).

I don't follow, is glibc missing entirely or will it be found once we install the right server arch? We don't plan to statically link the build, for now we just require that the remote meet these prereqs: https://code.visualstudio.com/docs/remote/linux#_remote-host-container-wsl-linux-prerequisites

**Comment by grebeplusplus** [comment]
> 
> 
> I guess to node, like https://github.com/nodejs/node, but I don't know, it may be upstream of them.
> 

Oh it's very much upstream from them. The only reason I knew what that error meant is that I've hit it before at my job.  That's the only reason I knew to use the file command to get the dependency list.

> > However you still have an unmet dependency in the case of the bug complaining that the system does not have glibc installed at all (for some reason I thought that was bug 443 but I seem to be mistaken I wish I could find the issue number that it was but it was late last night).
> 
> I don't follow, is glibc missing entirely or will it be found once we install the right server arch? We don't plan to statically link the build, for now we just require that the remote meet these prereqs: https://code.visualstudio.com/docs/remote/linux#_remote-host-container-wsl-linux-prerequisites

>'glibc >=2.17' 

Okay I didn't see that originally. If that's a stated dependency then that's fine. But it's still possible for someone to have an android or other linux base that uses LLMV based C runtimes exclusively. This would mean glibc is was never installed.

**Comment by grebeplusplus** [comment]
Just checking in to see if there were any updates on this?

**Comment by cakebake** [comment]
> Yeah currently I just look at the aarch64 in uname and assume the 64-bit version is what we want. So the right solution is just to also check `dpkg` and install the 32 bit version in this case right?

Thanks for the hint. My workaround is currently to boot the Pi with a 32 bit kernel, install the vscode-server (open remote in vscode) and then work with the 64 bit kernel again.

You can change this under /boot/config.txt.

Unfortunately, this has to be repeated with every vscode update.

Can you possibly include an option that the binaries are installed as `auto / 32- / 64-bit`?

**Comment by roblourens** [comment]
This came up again in https://github.com/microsoft/vscode-remote-release/issues/4558

Other people who have run into this, wondering if you can tell me the output of `getconf LONG_BIT` and 
```
file `which ls`
```

there is also still the option to just let users override this on a per-host basis, or to ask the user for the architecture when we detect it may be ambiguous, but I really want to find a nice portable automatic solution.

**Comment by mvadu** [comment]
@roblourens here you go.. I was about to open a new issue, and noticed this.

```shell
file `which ls`
/bin/ls: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, BuildID[sha1]=67a394390830ea3ab4e83b5811c66fea9784ee69, stripped

getconf LONG_BIT
32

uname -a
Linux raspberrypi4 5.10.17-v8+ #1403 SMP PREEMPT Mon Feb 22 11:37:54 GMT 2021 aarch64 GNU/Linux

```

**Comment by PavelSosin-320** [comment]
@mvadi After upgrading my distros and having troubles with the shutdown I found a new toy from the systemd brand:
hostnamectl that returns 
Static hostname: MSI-wsl
       Icon name: computer-container
         Chassis: container
      Machine ID: 051d2c9dfd574a228e61a4776f93258f
         Boot ID: 3c35823b30c84aadb66ffe762499f3f4
  Virtualization: wsl
Operating System: Fedora 34 (Container Image Prerelease)
     CPE OS Name: cpe:/o:fedoraproject:fedora:34
          **Kernel: Linux 5.4.91-microsoft-standard-WSL2
    Architecture: x86-64**
for my case and ability to ask remotely:
Options:
  -h --help              Show this help
     --version           Show package version
     --no-ask-password   Do not prompt for password
  -H --host=[USER@]HOST  Operate on remote host
  **-M --machine=CONTAINER Operate on local container**
Is it usable for VSCode-remote-release?

**Comment by nlsweet** [comment]
`hostnamectl` does return the correct architecture of the OS in certain cases.  In my case, we have a 64 bit capable ARM processor, but the installed OS is 32 bit. `hostnamectl` returns "arm64" (and `uname -m` returns aarch64), but 64 bit binaries fail to run and other techniques for determining the architecture (e.g., `getconf LONG_BIT` or `file /usr/bin/ls`) return 32 bit.

**Comment by mvadu** [comment]
Exactly.. I think the `getconf LONG_BIT` and/or `file <some userland executable>` are more reliable way to detect the platform.

**Comment by n8henrie** [comment]
Was excited to see VS Code Remote SSH supporting Raspberry Pi now! Did not last time I checked (a while back).

Unfortunately I ran into this immediately -- aarch64 Raspbian on RPi 4.

Tried `sudo apt install nodejs` and changing the path in `server.sh` to use `/usr/bin/node` instead of `./node` -- it gets a bit farther this way but then complaints about some ports. Tried disabling my firewall, and then it eventually ends up with `Error: WebSocket close with status code 1006)`.

¯\\_(ツ)_/¯

**Comment by grebeplusplus** [comment]
I'm surprised to see this still active almost a year later. With "official" support for raspberry pi now here can we drag this back out of the backlog graveyard and work on a fix? I'm happy to help where I can but we're out of my domain here. I'm just a dogged debugger :)

**Comment by zsol** [comment]
Here's a different workaround while we wait for a better fix (which would be for vscode to check `getconf LONG_BIT` instead of `uname -a`):

Put a file with the following contents somewhere in your `$PATH` before `/bin` (for example into `/usr/local/bin/uname`)
```
#!/bin/sh
exec /bin/uname "$@" | sed 's/aarch64/armv7l/'
```

Make sure the file is executable, then `rm -rf ~/.vscode-server` and try reconnecting.

**Comment by Omoeba** [comment]
Any progress on this?

**Comment by woaiwinnie2** [comment]
Same behavior with the official release of arm64 Raspbian.

**Comment by juan57** [comment]
I had the same issue - installing the `armv7l` arch of the vs remote server manually worked for me.

I modified and used the script from: https://gist.github.com/b01/0a16b6645ab7921b0910603dfb85e4fb
```
#!/bin/sh

set -e

# Auto-Get the latest commit sha via command line.
get_latest_release() {
    tag=$(curl --silent "https://api.github.com/repos/${1}/releases/latest" | # Get latest release from GitHub API
          grep '"tag_name":'                                              | # Get tag line
          sed -E 's/.*"([^"]+)".*/\1/'                                    ) # Pluck JSON value

    tag_data=$(curl --silent "https://api.github.com/repos/${1}/git/ref/tags/${tag}")

    sha=$(echo "${tag_data}"           | # Get latest release from GitHub API
          grep '"sha":'                | # Get tag line
          sed -E 's/.*"([^"]+)".*/\1/' ) # Pluck JSON value

    sha_type=$(echo "${tag_data}"           | # Get latest release from GitHub API
          grep '"type":'                    | # Get tag line
          sed -E 's/.*"([^"]+)".*/\1/'      ) # Pluck JSON value

    if [ "${sha_type}" != "commit" ]; then
        combo_sha=$(curl -s "https://api.github.com/repos/${1}/git/tags/${sha}" | # Get latest release from GitHub API
              grep '"sha":'                                                     | # Get tag line
              sed -E 's/.*"([^"]+)".*/\1/'                                      ) # Pluck JSON value

        # Remove the tag sha, leaving only the commit sha;
        # this won't work if there are ever more than 2 sha,
        # and use xargs to remove whitespace/newline.
        sha=$(echo "${combo_sha}" | sed -E "s/${sha}//" | xargs)
    fi

    printf "${sha}"
}

ARCH="x64"
U_NAME=$(uname -m)

# Raspberry pi 4b needs armv7l (armhf), but uname is specified as aarch64 (arm64) so I modified this part.
if [ "${U_NAME}" = "aarch64" ]; then
    ARCH="armhf"
fi

archive="vscode-server-linux-${ARCH}.tar.gz"
owner='microsoft'
repo='vscode'
commit_sha=$(get_latest_release "${owner}/${repo}")

if [ -n "${commit_sha}" ]; then
    echo "will attempt to download VS Code Server version = '${commit_sha}'"

    # Download VS Code Server tarball to tmp directory.
    curl -L "https://update.code.visualstudio.com/commit:${commit_sha}/server-linux-${ARCH}/stable" -o "/tmp/${archive}"

    # Make the parent directory where the server should live.
    # NOTE: Ensure VS Code will have read/write access; namely the user running VScode or container user.
    mkdir -vp ~/.vscode-server/bin/"${commit_sha}"

    # Extract the tarball to the right location.
    tar --no-same-owner -xzv --strip-components=1 -C ~/.vscode-server/bin/"${commit_sha}" -f "/tmp/${archive}"
else
    echo "could not pre install vscode server"
fi
```

So, what I did:
1. `rm -rf ~/.vscode-server` # remove existing vscode-server
2. ^create an sh file and copy paste the above script
3. run the script.
4. VS code remote should work

Just a guess but I figure is that for RPis that are showing as arch `aarch64`, downloads the `arm64` version when really, it should download the `armv7l` arch, which is `armhf`.

Hope this helps someone avoid 2 hours of headbashing :P

**Comment by Psycoguana** [comment]
Thank you so much @juan57 ! It worked like a charm

**Comment by roblourens** [comment]
Hi all, I've attempted a fix here. It will be out in the __pre-release__ version of the Remote-SSH extension on Monday morning. I'm including a .vsix here and would appreciate someone installing this manually and letting me know whether it works.

Download this, __rename__ to `.vsix`, and use the command "Install from VSIX" command to install

[remote-ssh-0.100.1.zip](https://github.com/microsoft/vscode-remote-release/files/11182025/remote-ssh-0.100.1.zip)

**Comment by ivanvoznyakovsky** [comment]
@roblourens new version works as expected. thanks a lot 💯

**Comment by okelet** [comment]
@roblourens working here also. Thanks!

**Comment by erik434** [comment]
@roblourens Thank you! Here's nother confirmation that your fix seems to resolve the problem. I installed the version you linked above, and I'm now able to develop remotely on my Raspberry Pi. My case seems similar to the others: 64-bit kernel but with 32-bit userspace installed. In case it's useful:

```
getconf LONG_BIT
32
```

```
dpkg --print-architecture
armhf
```

```
uname -a
Linux octopi 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux
```

```
file `which ls`
/usr/bin/ls: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, BuildID[sha1]=81004d065160807541b79235b23eea0e00a2d44e, for GNU/Linux 3.2.0, stripped
```

**Comment by bbqchef47** [comment]
Yes, this apparently worked as I followed the instructions and then deleted the existing vscode server  on the remote (rm -rf .vscode-server ). restarted VSCODE in windows and connected via Remote Explorer. Full functionality as far as I can tell...
I did ask my buddy Bard about this and it has no ideas about this other than referring to this series of links.
Thanks to whoever is involved in solving.
